vng/vng_test.go

```diff
@@ -0,0 +1,351 @@
+		panic("Unreachable")
+	case zed.TypeFloat32:
+		builder.Append(zed.EncodeFloat32(math.Float32frombits(binary.LittleEndian.Uint32(genBytes(b, 4)))))
+	case zed.TypeFloat64:
+		builder.Append(zed.EncodeFloat64(math.Float64frombits(binary.LittleEndian.Uint64(genBytes(b, 8)))))
+	case zed.TypeBool:
+		builder.Append(zed.EncodeBool(genByte(b) > 0))
+	case zed.TypeBytes:
+		builder.Append(zed.EncodeBytes(genBytes(b, int(genByte(b)))))
+	case zed.TypeString:
+		builder.Append(zed.EncodeString(string(genBytes(b, int(genByte(b))))))
+	case zed.TypeIP:
+		builder.Append(zed.EncodeIP(netip.AddrFrom16(*(*[16]byte)(genBytes(b, 16)))))
+	case zed.TypeNet:
+		ip := netip.AddrFrom16(*(*[16]byte)(genBytes(b, 16)))
+		numBits := int(genByte(b)) % ip.BitLen()
+		net, err := ip.Prefix(numBits)
+		if err != nil {
+			// Should be unreachable.
+			panic(err)
+		}
+		builder.Append(zed.EncodeNet(net))
+	case zed.TypeType:
+		typ := genType(b, context, 3)
+		builder.Append(zed.EncodeTypeValue(typ))
+	case zed.TypeNull:
+		builder.Append(nil)
+	default:
+		switch typ := typ.(type) {
+		case *zed.TypeRecord:
+			builder.BeginContainer()
+			for _, field := range typ.Fields {
+				genValue(b, context, field.Type, builder)
+			}
+			builder.EndContainer()
+		case *zed.TypeArray:
+			builder.BeginContainer()
+			for genByte(b) != 0 {
+				genValue(b, context, typ.Type, builder)
+			}
+			builder.EndContainer()
+		case *zed.TypeMap:
+			builder.BeginContainer()
+			for genByte(b) != 0 {
+				genValue(b, context, typ.KeyType, builder)
+				genValue(b, context, typ.ValType, builder)
+			}
+			builder.TransformContainer(zed.NormalizeMap)
+			builder.EndContainer()
+		case *zed.TypeSet:
+			builder.BeginContainer()
+			for genByte(b) != 0 {
+				genValue(b, context, typ.Type, builder)
+			}
+			builder.TransformContainer(zed.NormalizeSet)
+			builder.EndContainer()
+		case *zed.TypeUnion:
+			tag := binary.LittleEndian.Uint64(genBytes(b, 8)) % uint64(len(typ.Types))
+			builder.BeginContainer()
+			builder.Append(zed.EncodeInt(int64(tag)))
+			genValue(b, context, typ.Types[tag], builder)
+			builder.EndContainer()
+		default:
+			panic("Unreachable")
+		}
+	}
+}
+
+func genTypes(b *bytes.Reader, context *zed.Context, depth int) []zed.Type {
+	types := make([]zed.Type, 0)
+	for len(types) == 0 || genByte(b) != 0 {
+		types = append(types, genType(b, context, depth))
+	}
+	return types
+}
+
+func genType(b *bytes.Reader, context *zed.Context, depth int) zed.Type {
+	if depth < 0 || genByte(b)%2 == 0 {
+		switch genByte(b) % 19 {
+		case 0:
+			return zed.TypeUint8
+		case 1:
+			return zed.TypeUint16
+		case 2:
+			return zed.TypeUint32
+		case 3:
+			return zed.TypeUint64
+		case 4:
+			return zed.TypeInt8
+		case 5:
+			return zed.TypeInt16
+		case 6:
+			return zed.TypeInt32
+		case 7:
+			return zed.TypeInt64
+		case 8:
+			return zed.TypeDuration
+		case 9:
+			return zed.TypeTime
+		case 10:
+			// TODO Find a way to convert u16 to float16.
+			//return zed.TypeFloat16
```

Comment:
That'd be [float16.Frombits](https://pkg.go.dev/github.com/x448/float16#Frombits).